yarn-project/acir-simulator/src/client/client_execution_context.ts

```diff
@@ -171,15 +179,56 @@ export class ClientTxExecutionContext {
    * @param contractAddress - The contract address.
    * @param storageSlot - The storage slot.
    * @param preimage - new note preimage.
+   * @param nullifier - note nullifier
+   * @param innerNoteHash - inner note hash
    */
-  public async pushNewNote(contractAddress: AztecAddress, storageSlot: ACVMField, preimage: ACVMField[]) {
+  public async pushNewNote(contractAddress: AztecAddress, storageSlot: Fr, preimage: Fr[], innerNoteHash: Fr) {
     const wasm = await CircuitsWasm.get();
     const nonce = computeCommitmentNonce(wasm, this.txNullifier, this.pendingNotes.length);
     this.pendingNotes.push({
       contractAddress,
-      storageSlot: fromACVMField(storageSlot),
+      storageSlot: storageSlot,
       nonce,
-      preimage: preimage.map(f => fromACVMField(f)),
+      preimage,
+      innerNoteHash,
     });
   }
+
+  /**
+   * Adding a nullifier into the current set of all pending nullifiers created
+   * within the current transaction/execution.
+   * @param nullifier - The pending nullifier to add in the list.
+   */
+  public pushPendingNullifier(nullifier: Fr) {
+    this.pendingNullifiers.add(nullifier);
+  }
+
+  /**
+   * Update the list of pending notes by chopping a note which is nullified.
+   * The identifier used to determine matching is the inner note hash value.
+   * However, we adopt a defensive approach and ensure that the contract address
+   * and storage slot do match.
+   * Note that this method might be called with an innerNoteHash referring to
+   * a note created in a previous transaction which will result in this array
+   * of pending notes left unchanged.
+   * @param innerNoteHash - The inner note hash value to which note will be chopped.
+   * @param contractAddress - The contract address
+   * @param storageSlot - The storage slot as a Field Fr element
+   */
+  public nullifyPendingNotes(innerNoteHash: Fr, contractAddress: AztecAddress, storageSlot: Fr) {
+    // IMPORTANT: We do need an in-place array mutation of this.pendingNotes as this array is shared
+    // by reference among the nested calls. That is the main reason for 'splice' usage below.
+    this.pendingNotes.splice(
+      0,
+      this.pendingNotes.length,
+      ...this.pendingNotes.filter(
+        n =>
+          !(
+            n.innerNoteHash.equals(innerNoteHash) &&
+            n.contractAddress.equals(contractAddress) &&
+            n.storageSlot.equals(storageSlot)
+          ),
+      ),
+    );
+  }
```

Comment:
Nice! As a followup task, we can probably replace the PrivateFunctionExecution's newNotes and then clean up [this chunk of code in the kernel prover](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec-rpc/src/kernel_prover/kernel_prover.ts#L172)